### PR TITLE
[acme] configuration: enable/disable ACME service

### PR DESCRIPTION
**Pull request by frasertweedale**
Add a config knob to disable and enable the ACME service entirely.  (This is
needed for FreeIPA use case).  Also add interfaces for dynamically updating
this configuration, and a file-watching implementation (LDAP - also for FreeIPA
use case - will come later).

### CODE REVIEW

File: base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngine.java [hunk 1]
```diff
@@ -162,6 +165,56 @@ public void setIssuer(ACMEIssuer issuer) {
...
+    public boolean isEnabled() {
+        return this.enabled;
+    }
+
+    private void setEnabled(boolean b) {
+        this.enabled = b;
+    }
+
+    public void loadEngineConfig(String filename) throws Exception {
+        logger.info("Loading ACME engine config " + filename);
```
[the remark this answers is not in the document]
Comment (frasertweedale) [reply base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngine.java:180]: Yes, I'll add an example.

File: base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngine.java [hunk 2]
```diff
@@ -162,6 +165,56 @@ public void setIssuer(ACMEIssuer issuer) {
...
+        if (f.exists()) {
+            try (FileReader reader = new FileReader(f)) {
+                cfg.load(reader);
+                className = cfg.getProperty("class");
+            }
+        }
+
+        if (className == null) {
+            // the default class just sends the default config values
+            className = "org.dogtagpki.acme.server.ACMEEngineConfigDefaultSource";
```
[the remark this answers is not in the document]
Comment (frasertweedale) [reply base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngine.java:195]: Aha, fair call, I'll update it.

File: base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngineConfigFileSource.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        filename = cfg.getProperty("filename");
+        if (null == filename) {
+            throw new IllegalArgumentException("filename must not be null");
+        }
+
+        // initial load
+        loadFile();
+
+        // start file monitor thread
+        (new Thread(this, "ACMEEngineConfigFileSource")).start();
```
[the remark this answers is not in the document]
Comment (frasertweedale) [reply base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngineConfigFileSource.java:46]: OK thanks for the pointer, I'll find a way to do it.

File: base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngineConfigFileSource.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                if (!wk.reset()) {
+                    // watch key could not be reset; break loop
+                    break;
+                }
+            }
+        } catch (Throwable e) {
+            logger.error(
+                "ACMEEngineConfigFileSource: caught exception while monitoring file",
+                e
+            );
```
[the remark this answers is not in the document]
Comment (frasertweedale) [reply base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngineConfigFileSource.java:92]: I'm happy to have it just stop.

**Comment by frasertweedale** [comment]
> FYI, right now it's already possible to enable/disable ACME using the `pki-server acme-deploy` or `acme-undeploy` commands. These commands basically will create/remove the deployment descriptor file for ACME webapp in Tomcat. Tomcat will monitor that file and will deploy/undeploy the webapp automatically in a few seconds.
> 
> The above mechanism can be used to change ACME configuration (e.g. `database.conf`). When the webapp is redeployed, it will reinitialize the ACME engine with the new configuration.
> 
> This PR is adding an `engine.conf` to enable/disable the ACME engine dynamically, but IIUC it doesn't actually initialize/shutdown the resources used by the engine. How about we use this mechanism to monitor changes in other ACME configuration (e.g. `database.conf`) instead and reinitialize the engine dynamically without deploy/undeploy?

Right, so this is "phase one" - support for dynamic reconfiguration of the other parts of the engine (well, some of them - the Issuer and challenge configuration at least) will follow.

The deploy/undeploy approach is not adequate for the IPA use case, where enable/disable/reconfiguration needs to happen *deployment-wide* based on change to LDAP object(s).  Obviously the LDAP configuration provider is not implemented in this PR, but the reconfiguration mechanisms are, and the file monitoring provider can be used to verify the behaviour and the adequacy of the interfaces.

**Comment by frasertweedale** [comment]
@edewata thanks for reviewing.  I think I have addressed your concerns in latest patchset.

**Comment by edewata** [comment]
@frasertweedale Any update on this PR?
@cipherboy Could you take a look at this as well?

**Comment by edewata** [comment]
@frasertweedale Thanks for the update, but I don't see any change regarding to separating the monitor configuration (i.e. monitor.conf). Are you still working on it, or do you want to merge it as is and then we refactor it later?

**Comment by frasertweedale** [comment]
@edewata the last update was just a trivial rebase.

I think we can merge it.  I agree the current example files are a bit confusing but we can tidy it up later.  The implementation does already support using separate files.  I'm comfortable with `engine.conf` as filename for monitor configuration; in the cleanup we could e.g. use `service.conf` for the other file.

The other issue was not restarting the file monitoring if there's a problem.  I can live with that, unless it's shown to be an issue under normal operating conditions.  Fixing it is a small change but it's not a high priority (for me).

**Comment by edewata** [comment]
How does the `service.conf` look like? ACME has serveral configuration files, and we probably want to monitor all of them:
* engine.conf
* metadata.conf
* database.conf
* issuer.conf
* validators.conf

So I don't think it makes sense to define one monitor in `engine.conf` and the rest in `service.conf`. I think it will make more sense to define all monitors in a single `monitors.conf` as I described above. That will keep the `engine.conf` clean.

**Comment by frasertweedale** [comment]
@edewata I don't think we need to monitor all of them - at least not right away.  My work is driven by what is needed for the IPA use case and I intend to do what is required for that, and leave the rest (for now).  So I agree on this: the approach should let us add monitors for more parts of the configuration as and when we need to.  So I will make a single `configsources.conf` and all monitors can be configured there.

File: base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngine.java [hunk 5]
```diff
@@ -162,6 +167,47 @@ public void setIssuer(ACMEIssuer issuer) {
...
+        // the configuration (Consumer<T>).  This abstraction ensures the
+        // ConfigSource has no direct access to the ACMEEngine instance.
+        engineConfigSource.init(
+            cfg,
+            new Consumer<Boolean>() {
+                @Override public void accept(Boolean b) {
+                    setEnabled(b);
+                    logger.info(
+                        "ACME service is "
+                        + (b ? "enabled" : "DISABLED")
```
Comment (cipherboy) [inline base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngine.java:203]: nit: same casing?

File: base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngineConfigFileSource.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (cipherboy) [reply base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngineConfigFileSource.java:111]: I'd prefer to have it monitor the file again, if it exists, perhaps with an exponential backout. But we can change behavior later.

File: base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngine.java (hunk 5 above)
Comment (frasertweedale) [reply base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngine.java:203]: I wanted to make DISABLED stand out a bit more in the log :)

**Comment by frasertweedale** [comment]
@edewata @cipherboy thanks for the reviews.  I think this PR ended up in a much better place than where it started.

**Pull request merged**